tools/clang/include/clang/AST/HlslTypes.h

```diff
@@ -215,35 +217,22 @@ struct UnusualAnnotation {
 struct RegisterAssignment : public UnusualAnnotation
 {
   /// <summary>Initializes a new RegisterAssignment in invalid state.</summary>
-  RegisterAssignment() : UnusualAnnotation(UA_RegisterAssignment),
-    ShaderProfile(), IsValid(false),
-    RegisterType(0), RegisterNumber(0), RegisterSpace(0), RegisterOffset(0)
-  {
-  }
-
-  RegisterAssignment(const RegisterAssignment& other) : UnusualAnnotation(UA_RegisterAssignment, other.Loc),
-    ShaderProfile(other.ShaderProfile),
-    IsValid(other.IsValid),
-    RegisterType(other.RegisterType),
-    RegisterNumber(other.RegisterNumber),
-    RegisterSpace(other.RegisterSpace),
-    RegisterOffset(other.RegisterOffset)
-  {
-  }
+  RegisterAssignment() : UnusualAnnotation(UA_RegisterAssignment) { }
 
   llvm::StringRef   ShaderProfile;
-  bool              IsValid;
-  char              RegisterType; // 'x' means only space is assigned from the source code
-  uint32_t          RegisterNumber;
-  uint32_t          RegisterSpace;
-  uint32_t          RegisterOffset;
+  bool              IsValid = false;
+  char              RegisterType = 0; // Lower-case letter, 0 if not explicitly set
+  uint32_t          RegisterNumber = 0; // Iff RegisterType != 0
+  llvm::Optional<uint32_t> RegisterSpace; // Set only if explicit "spaceN" syntax
+  uint32_t          RegisterOffset = 0;
 
   void setIsValid(bool value) {
     IsValid = value;
   }
 
-  void setAsSpaceOnly() { RegisterType = 'x'; }
-  bool isSpaceOnly() const { return RegisterType == 'x'; }
+  bool isSpaceOnly() const {
+    return RegisterType == 0 && RegisterSpace.hasValue();
```

Comment:
Actually, I wonder if `llvm::Optional` is necessary?

IIUC the absence of `space#` implies that space #0 should be used, so with `llvm::Optional`, we will end up using `RegisterSpace.getValueOr(0)` everywhere. Without `llvm::Optional`, we can initialize `RegisterSpace` to zero; only set its value when provided in the HLSL source; and use `RegisterSpace` as a uint everywhere.

And `isSpaceOnly` function can just `return RegisterType == 0;`

What do you think?

Reply:
(as a side: the fix for the SPIR-V backend is just a matter of adding `getValueOr` in some places I believe, and I wouldn't mind doing that if you think `llvm::Optional` makes more sense)

Reply:
We have an `-auto-binding-space` parameter to choose the space for resources with no explicit assignment, which means that I must distinguish between `Buffer buf;` and `Buffer buf : register(space0);` because it would only apply to the former. I guess I could distinguish based on the presence of the RegisterAllocation UnusualAttribute, but since its space parameter is optional, the current data model maps better to the syntax.

Reply:
Ah! I see. Sorry I forgot about `-auto-binding-space`. Makes sense. I'll update the backend code. Thanks!